snmp/datadog_checks/snmp/models.py

```diff
@@ -23,46 +24,29 @@ class OID(object):
     def __init__(self, value):
         # type: (Union[Sequence[int], str, ObjectName, ObjectIdentity, ObjectType]) -> None
         parts = None  # type: Optional[Tuple[int, ...]]
+        object_identity = None  # type: Optional[ObjectIdentity]
 
         try:
             parts = parse_as_oid_tuple(value)
         except CouldNotDecodeOID:
             raise  # Invalid input.
         except UnresolvedOID:
+            # `value` must be a PySNMP object, let's inspect it.
             if isinstance(value, ObjectType):
-                # An unresolved `ObjectType(ObjectIdentity('<MIB>', '<symbol>'))`.
-                parts = None
+                object_identity = object_identity_from_object_type(value)
             elif isinstance(value, ObjectIdentity):
-                # An unresolved `ObjectIdentity('<MIB>', '<symbol>')`.
-                parts = None
+                object_identity = value
             else:  # pragma: no cover
                 raise RuntimeError('Unexpectedly treated {!r} as an unresolved OID'.format(value))
-
-        # Resolve the `ObjectIdentity` which we can use to resolve the MIB name of the OID (for metric naming).
-        # PySNMP objects may contain MIB information already, so check for them in priority.
-        if isinstance(value, ObjectType):
```

Comment:
If seems a bit more than refactoring. Just to confirm my understanding:

1. we relied on this removed code before to detect the `object_identity` using `isinstance(value, ObjectType)` and `isinstance(value, ObjectIdentity)`:

https://github.com/DataDog/integrations-core/blob/master/snmp/datadog_checks/snmp/models.py#L43-L49

But now we only detect the `object_identity` using `isinstance(value, ObjectType)` and `isinstance(value, ObjectIdentity)` IF `parts = parse_as_oid_tuple(value)` fail (before we always execute the code above).

2.  Before, `parts` is None if `parts = parse_as_oid_tuple(value)` (line 30).

We now also derived `parts` from `object_identity`: `parts = parts_from_object_identity(object_identity)`

Reply:
Yes, you're right, the approach has changed a bit.

1. I think this is an oversight from me. The comment on this line is correct: if `parse_as_oid_tuple` succeeds we should still use the provided `ObjectType`/`ObjectIdentity` as otherwise we may lose MIB information (eg `IF-MIB::inInErrors`) and won't be able to resolve later.
2. This is a "new feature": it's actually possible to derive `parts` even if the `ObjectIdentity` is not resolved yet, in the case when it was built as `ObjectIdentity('1.2.3.4.5...')`. I should probably keep this for later.

At this point I think expanding the usage of `OID` across the SNMP check (esp in `OIDResolver`) to see what actual problems we may have for either approaches is a prerequisite...

Reply:
Thx for the explanation.

As general rule, I think it's better to separate refactoring and behaviour/feature change into individual PRs.

> Probably rework the PR to remove any behavior changes, and stick to adding the pysnmp_inspect.py module.

👍